internal/runbits/runtime/progress/decor.go

```diff
@@ -0,0 +1,166 @@
+package progress
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/ActiveState/cli/internal/constants"
+	"github.com/ActiveState/cli/internal/errs"
+	"github.com/ActiveState/cli/internal/locale"
+	"github.com/ActiveState/cli/internal/logging"
+	"github.com/ActiveState/cli/internal/termutils"
+	"github.com/ActiveState/cli/pkg/platform/runtime/artifact"
+	"github.com/vbauerster/mpb/v7"
+	"github.com/vbauerster/mpb/v7/decor"
+)
+
+const progressBarWidth = 40
+
+var spinnerFrames = []string{`|`, `/`, `-`, `\`}
+
+var refreshRate = constants.TerminalAnimationInterval
+
+type bar struct {
+	*mpb.Bar
+	started time.Time
+	total   int64
+}
+
+func (b *bar) setInternalTotal(v int64) {
+	b.Bar.SetTotal(v, false)
+	b.total = v
+}
+
+// Completed reports whether the bar has reached 100%. We have our own assertion prior to the mpb one as for whatever
+// reason mpb reports completed even when it isn't, and I've not been able to diagnose why.
+func (b *bar) Completed() bool {
+	if b.Bar.Current() < b.total {
+		return false
+	}
+
+	return b.Bar.Completed()
+}
+
+// trimName ensures that the name in a progress bar is not too wide for a terminal to display
+func (p *ProgressDigester) trimName(name string) string {
+	if len(name) > p.maxNameWidth {
+		return name[0:p.maxNameWidth]
+	}
+	return name
+}
+
+// addTotalBar adds a bar counting a number of sub-events adding up to total
+func (p *ProgressDigester) addTotalBar(name string, total int64, options ...mpb.BarOption) *bar {
+	logging.Debug("Adding total bar: %s", name)
+	return p.addBar(name, total, false, append(options, mpb.BarFillerClearOnComplete())...)
+}
+
+// addSpinnerBar adds a bar with a spinning progress indicator
+func (p *ProgressDigester) addSpinnerBar(name string, options ...mpb.BarOption) *bar {
+	logging.Debug("Adding spinner bar: %s", name)
+	return &bar{
+		p.mainProgress.Add(1,
+			mpb.NewBarFiller(mpb.SpinnerStyle(spinnerFrames...)),
+			append(options,
+				mpb.BarFillerClearOnComplete(),
+				mpb.PrependDecorators(
+					decor.Name(name, decor.WC{W: p.maxNameWidth, C: decor.DidentRight}),
+				),
+				mpb.AppendDecorators(
+					decor.OnComplete(decor.NewPercentage("", decor.WC{W: 5}), ""),
+				),
+			)...,
+		), time.Now(), 1,
```

Comment:
nit: just flagging this unkeyed type initialization

Reply:
This one seems fine to me. There's no rule that all structs should be keyed. Just that in the previous ones I addressed there was a lot being initialized and the keys were important for reading the code. I don't think the same applies here.